Implementing the firebase storage and getting rid of the prefix "checkit_"

andrayantelo opened the pull request:
Made a firebase storage object. Resolves #110 and #112. It gets initialized in the app method initFirebase and get passed auth. Has a few of the same methods as the localStorage store object.  Started the getAllCalendarIds method in the firebase storage object.  The firebase database once method doesn't do startWork and endWork like my methods do in the localStorage store object so I'm not sure how to use startWork and endWork in the getAllCalendarIds method yet.

Still WIP because I want to merge the new database rules into it, which means I will have to restructure how saves are made.

andrayantelo committed:
Made an object for the firebase storage
It gets initialized in the app method initFirebase and get
passed auth.
Has a few of the same methods as the localStorage store object.
Started the getAllCalendarIds method in the firebase storage object.
The firebase database once method doesn't do startWork and endWork
like my methods do in the localStorage store object so I'm not sure how
to use startWork and endWork in the getAllCalendarIds method yet.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -865,6 +870,95 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+        // Returns a promise for the allCalendarIds object from storage
+        
+        var userId = self.user.currentUser.uid;
+        self.database.ref('/users/' + userId + '/checkit_allCalendarIds')
```
This is the promise that you return. Call `startWork` before it starts and call `endWork` in the handlers instead of just logging.

andrayantelo committed:
Fixed getAllCalendarIds so that it returns promise

andrayantelo committed:
Firebase store object save method
Right now it just saves the allCalendarIds for the current user. It also needs to store the calendar states.

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -923,17 +923,58 @@ var firebaseCalendarStorage = function(user) {
+        //save an App object (like a calendar object for example) in storage
+        
+        // Store the calendar state in the firebase database.
+        
```
@johntyree Do you think I should just have an objects in the /users/uid/ path like users/uid/calendars
and in there have an object like this:
{ calendar_uniqueId: { state object}}

Instead of the way it is now?

andrayantelo replied:
Maybe users/uid/calUniqueId/calStateObject

johntyree replied:
I think it will make things like sharing harder later if every calendar is "part of" a user, but if you're having trouble getting it to work at all like it is now, or you have a better reason to do it, then go for it.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -865,6 +870,116 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+        
+            self.getAllCalendarIds()
+                .then(function (allCalendarIds) {
+                    console.log("setting new value in database");
```
log messages like this are a lot more useful if you include the relevant data

johntyree commented on public/calendar_reboot.js:
```diff
@@ -865,6 +870,116 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+                    firebase.database().ref().update(updates);
+                    console.log("DATA SET");
+                    self.endWork();
+                    resolve();  // At what point do I reject?
```
> ` // At what point do I reject?`

You reject when there's an error. In this case, the only error that might happen will come from the firebase library. What happens when there's a firebase error? Which code gets executed?

andrayantelo replied:
the code inside of the catch function right?

andrayantelo replied:
Pls respond @johntyree :cat:

andrayantelo replied:
I don't need a reject then in this case?

andrayantelo committed:
Update sample_calendar.json

Is this a good way to structure it to manage authorization for reading and writing to different calendars?

andrayantelo committed:
Update database.rules.json

Not sure if this is correct.

andrayantelo committed:
Added methods to firebase store object
Have not written them for firebase yet. Added a few comments.

andrayantelo committed:
Edited JSON file for valid JSON

andrayantelo committed:
Fixed syntax errors in rules file

andrayantelo committed:
Figuring out how to write the rules

andrayantelo committed:
Updated the save method in the firebase store
Firebase store now stores the calendarState in calendars/calUniqueID, updates the user's allCalendarIds, and updates the permissions for the newly stored calendar. Still unsure about how to use reject inside of the promise if I even need to.

andrayantelo committed:
Added remove methods to firebase store object

andrayantelo committed:
Added all necessary methods to firebase store obj
For some reason the user is undefined.

andrayantelo committed:
Added .val() where needed inside loadById
displayActiveCalendar appears to be running both then and catch.

andrayantelo committed:
Added a return to loadById method of firebaseStore

andrayantelo committed:
In attachClickHandler check if CheckedDays exists
Can't have something in the database point to an empty object. Maybe I can have checkedDays point to something like null?
Anyway, before attaching click handler we check if CheckedDays exists and if it doesn't we set it equal to an empty object.

andrayantelo committed:
Use firebase store obj save in attachClickHandler
Everything appears to be working and have changed all storage functions from store object to firebaseStore object.

andrayantelo committed:
Commented out old store object

andrayantelo committed:
Got rid of localStorage store object
Deleted some print statements as well

andrayantelo committed:
Fixed sample database json and got rid of prefix

andrayantelo committed:
Updated sample database JSON and database rules
Still need to write rules for the read and write of the "read" and "write" nodes inside of calendars/$calendarId/

andrayantelo committed:
Added nodes for write and read

johntyree commented on public/calendar_reboot.js:
```diff
@@ -136,20 +130,20 @@ CheckIt.prototype.displayActiveCalendar = function() {
-                   this.store.removeActive();
+               .catch(function(err) {
+                   console.error("Calendar does not exist " + err);
+                   this.firebaseStore.removeActive();  // Should I be doing this here?
```
It depends on the error probably. If the error is that the calendar doesn't exist then yes. If the error is because firebase couldn't load the data or the network was crappy and lost your request or something, then no.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
+        var updates = {};
+        updates['users/' + userId + '/allCalendarIds/' + calUniqueId] = calTitle;
+        updates['calendars/' + calUniqueId + '/calendarState/'] = calState;
+        updates['calendars/' + calUniqueId + '/permissionRead/'] = userId;
```
Remember to update this based on #116

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
+        return new Promise(function(resolve, reject) {
+            self.database.ref().update(updates);
+            self.endWork();
+            resolve();  // Still unsure whether I will need to use reject in here or not.
```
The call to `update()` above is asynchronous...

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
+        
+        self.startWork();
+        return new Promise( function(resolve, reject) {
+            self.database.ref().update(updates);
```
Again this is an asynchronous call, so this whole function isn't going to behave the way you want.
https://firebase.google.com/docs/reference/js/firebase.database.Reference#update

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
+        return new Promise( function(resolve, reject) {
+            self.database.ref().update(updates);
+            self.endWork();
+            resolve();  // Reject somewhere?
```
> `// Reject somewhere?`

You want to `reject()` iff there's an error that the caller could `.catch()` and do something with. What might be an error here? Where would it come from?

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
-            .catch(function () {
-                console.log("Unable to remove calendar");
-            });
+        // Remove a calendar from storage by using it's Id.
```
its

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
+        // First deleting it from /user/uid/allCalendarIds
+        // Second remove permissions
+        // third removing it from calendars/
+        // removing currentActive status has it's own method
```
its

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
+        var userId = self.user.uid;
         
+        self.startWork();
+        return new Promise(function(resolve, reject) {
```
This is all broken because the `.remove()` calls are asynchronous. What is going to be the order of events in this function?

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,168 @@ var LocalCalendarStorage = function(params) {
-                jitter(resolve, calendar);
-            }
-            else jitter(reject, "Calendar not found");
+            self.database.ref('users/' + userId + '/currentActiveCalendar/').remove();
```
Async...

johntyree reviewed (changes requested):
After #116 lands, merge master back into this branch to get the updated rules and database files. Then make sure all your paths still make sense.

andrayantelo committed:
fixed read and write rules for calendars/$calendarId

andrayantelo committed:
removed a comment

andrayantelo committed:
updated rules for the read and write permissions. Fixed JSON

andrayantelo committed:
Merge branch 'firebaseStorage' into database_structure

andrayantelo committed:
Merge pull request #116 from andrayantelo/database_structure

[WIP] Update sample_calendar.json and database.rules.json

andrayantelo committed:
Fixed asyn calls to database

johntyree commented on database.rules.json:
```diff
@@ -1,6 +1,26 @@
+          ".read" : "auth !== null && data.parent().child('read/' + auth.uid).exists()",
+          ".write" : "auth !== null && data.parent().child('write/' + auth.uid).exists()"
+        },
+        "write" : {
```
I feel like "write" and "read" here could have better names. Maybe `writers` and `readers` or `can_write` and `can_read`.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -136,20 +130,20 @@ CheckIt.prototype.displayActiveCalendar = function() {
-                   console.error("Calendar does not exist");
-                   this.store.removeActive();
+               .catch(function(err) {
+                   console.error("Calendar does not exist " + err);
```
Are you certain this means that the calendar doesn't exist? What if firebase itself just had a temporary problem?

johntyree commented on public/calendar_reboot.js:
```diff
@@ -865,24 +874,29 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
         
-        var key = prefix + "_" + id;
-        return key;
+        var randomNumber = Math.random() * 4000;
```
You probably don't want a 4s delay in master.

andrayantelo replied:
Do I even need jitter anymore?

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,187 @@ var LocalCalendarStorage = function(params) {
-            jitter(resolve);
-        });
+        // Store the calendar state in the firebase database.
+        // TO DO save the actual calendar state.
```
TODO not "TO DO"

Also isn't this being done?

andrayantelo replied:
Forgot to delete that comment.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,187 @@ var LocalCalendarStorage = function(params) {
         
-        //put calendar in allCalendarIdss and store it
+        // Store the user's allCalendarIds and their calendarState .. setting the
+        // active calendar has it's own method.
```
its

Also I don't understand these two comments. Did you accidentally

andrayantelo replied:
You don't have to save the currentActiveCalendar in the save method because setActive is its own method. I was just writing that there so that I could remember that.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,187 @@ var LocalCalendarStorage = function(params) {
     self.remove = function(calendarObj) {
-        //remove an app object (like a calendar) from storage
+        // Remove an app object (like a calendar) from storage
+        // To do: Determine if this method is necessary
```
TODO

johntyree commented on public/calendar_reboot.js:
```diff
@@ -921,151 +933,187 @@ var LocalCalendarStorage = function(params) {
-        //remove a calendar from storage by using it's Id.
+        // Remove a calendar from storage by using its Id.
+        // First deleting it from /user/uid/allCalendarIds
+        // Second remove permissions
```
This doesn't seem accurate.

andrayantelo committed:
Got rid of some comments, fixed other comments

johntyree commented on public/calendar_reboot.js:
```diff
@@ -136,20 +130,20 @@ CheckIt.prototype.displayActiveCalendar = function() {
-                   console.error("Calendar does not exist");
-                   this.store.removeActive();
+               .catch(function(err) {
+                   console.error("Could not load calendar " + err);
```
extra space between "not" and "load" :smiling_imp:

andrayantelo replied:
No there isn't :angel:

johntyree commented on public/calendar_reboot.js:
```diff
@@ -136,20 +130,20 @@ CheckIt.prototype.displayActiveCalendar = function() {
-                   this.store.removeActive();
+               .catch(function(err) {
+                   console.error("Could not load calendar " + err);
+                   this.firebaseStore.removeActive(); 
```
trailing whitespace

andrayantelo committed:
fixed a trailing white space issue
